src/v/datalake/tests/parquet_writer_test.cc

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "bytes/bytes.h"
+#include "datalake/arrow_translator.h"
+#include "datalake/parquet_writer.h"
+#include "datalake/tests/test_data.h"
+#include "iceberg/tests/value_generator.h"
+#include "utils/file_io.h"
+
+#include <arrow/array.h>
+#include <arrow/io/interfaces.h>
+#include <arrow/io/memory.h>
+#include <arrow/table.h>
+#include <arrow/type_fwd.h>
+#include <gtest/gtest.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/type_fwd.h>
+
+TEST(ParquetWriter, CreatesValidParquetData) {
+    datalake::arrow_translator schema_translator(
+      test_schema(iceberg::field_required::no));
+
+    iobuf full_result;
+
+    for (int i = 0; i < 5; i++) {
+        auto data = iceberg::tests::make_value(
+          iceberg::tests::value_spec{
+            .forced_fixed_val = iobuf::from("Hello world")},
+          test_schema(iceberg::field_required::no));
+        schema_translator.add_data(std::move(data));
+    }
+
+    std::shared_ptr<arrow::Array> result = schema_translator.take_chunk();
+    ASSERT_NE(result, nullptr);
+
+    datalake::arrow_to_iobuf writer(*schema_translator.build_arrow_schema());
+
+    for (int i = 0; i < 10; i++) {
+        writer.add_arrow_array(result);
+        iobuf serialized = writer.take_iobuf();
```

Comment:
it looks like there is a use-after-move issue here. when you call `writer.take_iobuf` that proxies to `iobuf iobuf_output_stream::take_iobuf() { return std::move(_current_iobuf); }`, but `iobuf` doesn't formally specify its moved-from-state. it works because it happens to leave it empty, but i think we should not depend on this. if you r-value qualified `iobuf_output_stream::take_iobuf()` per my other comment, then I think you'd have a cascade of changes which resulted in a use-after-move clang-tidy warning here.

i see two options. don't re-use the writer. this is the "cleanest" option, and if the writer is lightweight (it looks like it) then it probably makes the most sense.

the other option would be to not r-value qualify take_iobuf, and call it something like `iobuf reset()` which would move the current iobuf out of the ostream, and explicitly reset it.

Reply:
The writer includes a parquet::arrow::FileWriter which is stateful, so I don't think it would work create a new writer, but resetting the iobuf in take_iobuf should work.

Reply:
I made a copy of the iobuf so I could reset the _current_iobuf and return the copy. How can I return that by rvalue reference when it is allocated on the stack?

Reply:
>I made a copy of the iobuf so I could reset the _current_iobuf and return the copy. How can I return that by rvalue reference when it is allocated on the stack?

Does it work to do something like

```
iobuf take_iobuf() {
    iobuf b = std::move(_current_iobuf);
    _current_iobuf = {}; // or like _current_iobuf.clear() or somesuch
    return b;
}
```

Reply:
```
iobuf take_iobuf() { return std::exchange(_current_iobuf, {}); }
```

Reply:
+1 to tyler and andrew neither of which make a copy